Retain meta data set on importer nodes

tamask opened the pull request:
During the import process, ImporterMeshInstance3D nodes are replaced with MeshInstance3D nodes. This change copies over any meta data that might have been set during the import process, for example, in a GLTFDocumentExtension.

This is a resolution for https://github.com/godotengine/godot-proposals/discussions/6586.

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

AThousandShips commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -2168,6 +2168,16 @@ Node *ResourceImporterScene::_generate_meshes(Node *p_node, const Dictionary &p_
+		{
+			List<StringName> meta_list;
+			p_node->get_meta_list(&meta_list);
+			for (List<StringName>::Element *element = meta_list.front(); element; element = element->next()) {
```
```suggestion
			for (List<StringName>::Element *E = meta_list.front(); E; E = E->next()) {
```
We normally name it so

tamask replied:
Ah ok, great!

AThousandShips commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -2168,6 +2168,16 @@ Node *ResourceImporterScene::_generate_meshes(Node *p_node, const Dictionary &p_
+			List<StringName> meta_list;
+			p_node->get_meta_list(&meta_list);
+			for (List<StringName>::Element *E = meta_list.front(); E; E = E->next()) {
+				StringName meta_key = element->get();
```
```suggestion
				StringName meta_key = E->get();
```
This one too, sorry wasn't by my computer or I'd have made it a single suggestion but forgot

tamask commented:
I've updated this pull request to be more thorough with copying over meta data on various imported nodes, not just ImporterMeshInstance3D.

AThousandShips commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -2199,6 +2211,16 @@ void ResourceImporterScene::_add_shapes(Node *p_node, const Vector<Ref<Shape3D>>
+void ResourceImporterScene::_copy_meta(Node *p_src_node, Node *p_dst_node) {
+	List<StringName> meta_list;
+	p_src_node->get_meta_list(&meta_list);
+	for (List<StringName>::Element *E = meta_list.front(); E; E = E->next()) {
+		StringName meta_key = E->get();
```
```suggestion
	for (const StringName &meta_key : meta_list) {
```
More convenient approach thanks to iterators

tamask replied:
ah, nice, thanks!

tamask committed:
Retain meta data set on imported nodes

During the import process, many importer nodes are replaced with their
engine node counterparts. For example, ImporterMeshInstance3D is
replaced with a MeshInstance3D node. Any meta data set on these
importer nodes, i.e. through a GLTFDocumentExtension, are lost during
the conversion. This change copies over any meta data set on these
importer nodes to their engine counterparts.

tamask commented:
I've updated the pull request to also copy over metadata from ImporterMesh to ArrayMesh. In addition to ImporterMeshInstance3D, ImporterMesh is the other intermediary class used where metadata set in a GLTFDocumentExtension would be lost. 

Here's an example use case in a GLTFDocumentExtension that copies over the `extras` from the gltf json so that it's available in the packed scene in godot:

```gdscript
func _apply_extras_to_meshes(state: GLTFState) -> void:
    for index: int in range(state.meshes.size()):
        var extras: Dictionary = state.json.meshes[index].get("extras", {})
        var mesh: GLTFMesh = state.meshes[index]
        _apply_extras_to_meta(state, mesh.mesh, extras)
```

This way, custom properties set in blender on the mesh data can be conveyed over to the godot side via gltf.

lyuma commented:
Looks good and easily digestible so I think it should be safe in 4.3 - Would it be possible to implement the same change in `modules/gltf/extensions/gltf_document_extension_convert_importer_mesh.cpp` ?
Basically, the GLTFDocumentExtension is the corresponding process for runtime GLTFDocument import, since the editor import pipeline code is not enabled at runtime.

https://github.com/godotengine/godot/blob/master/modules/gltf/extensions/gltf_document_extension_convert_importer_mesh.cpp
It's only in one place, so duplicating the contents of `_copy_meta` is acceptable in this case.

tamask commented:
> Looks good and easily digestible so I think it should be safe in 4.3 - Would it be possible to implement the same change in `modules/gltf/extensions/gltf_document_extension_convert_importer_mesh.cpp` ? Basically, the GLTFDocumentExtension is the corresponding process for runtime GLTFDocument import, since the editor import pipeline code is not enabled at runtime.
> 
> https://github.com/godotengine/godot/blob/master/modules/gltf/extensions/gltf_document_extension_convert_importer_mesh.cpp It's only in one place, so duplicating the contents of `_copy_meta` is acceptable in this case.

Absolutely, I've updated the pull request source branch to reflect this.

lyuma reviewed (approved):
This will fix a source of confusion when working with import plugins. Great work!

lyuma commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -2553,6 +2563,7 @@ Node *ResourceImporterScene::_generate_meshes(Node *p_node, const Dictionary &p_
 			}
 
 			if (mesh.is_valid()) {
+				_copy_meta(*(src_mesh_node->get_mesh()), *mesh);
```
This suggestion seems like it is mixing two separate conventions. Would it not logically follow that all conversions from Ref<Foo> to Foo* should use .ptr(), not only the first one?
```suggestion
				_copy_meta(src_mesh_node->get_mesh().ptr(), mesh.ptr());
```
Actually come to think of it, usually in C++, operator* returns Foo&, but it seems Ref<Foo>::operator* returns Foo*?

tamask replied:
I'll update both arguments to use the `.ptr()` convention

tamask replied:
Updated the source branch to reflect this suggestion:

https://github.com/cardboardcomputer/godot/blob/e19f0042caddb7d0ceb260e9ba9f5e5ee08a9f44/editor/import/3d/resource_importer_scene.cpp#L2566

https://github.com/cardboardcomputer/godot/blob/e19f0042caddb7d0ceb260e9ba9f5e5ee08a9f44/modules/gltf/extensions/gltf_document_extension_convert_importer_mesh.cpp#L71

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

tamask commented:
> Thanks! And congrats for your first merged Godot contribution 🎉

Wonderful, and thank you!
